superset/dashboards/permalink/api.py

```diff
@@ -0,0 +1,171 @@
+
+from superset.constants import MODEL_API_RW_METHOD_PERMISSION_MAP, RouteMethod
+from superset.dashboards.commands.exceptions import (
+    DashboardAccessDeniedError,
+    DashboardNotFoundError,
+)
+from superset.dashboards.permalink.commands.create import (
+    CreateDashboardPermalinkCommand,
+)
+from superset.dashboards.permalink.commands.get import GetDashboardPermalinkCommand
+from superset.dashboards.permalink.exceptions import DashboardPermalinkInvalidStateError
+from superset.dashboards.permalink.schemas import DashboardPermalinkPostSchema
+from superset.extensions import event_logger
+from superset.key_value.exceptions import KeyValueAccessDeniedError
+from superset.views.base_api import requires_json
+
+logger = logging.getLogger(__name__)
+
+
+class DashboardPermalinkRestApi(BaseApi):
+    add_model_schema = DashboardPermalinkPostSchema()
+    method_permission_name = MODEL_API_RW_METHOD_PERMISSION_MAP
+    include_route_methods = {
+        RouteMethod.POST,
+        RouteMethod.PUT,
+        RouteMethod.GET,
+        RouteMethod.DELETE,
+    }
+    allow_browser_login = True
+    class_permission_name = "DashboardPermalinkRestApi"
+    resource_name = "dashboard"
+    openapi_spec_tag = "Dashboard Permanent Link"
+    openapi_spec_component_schemas = (DashboardPermalinkPostSchema,)
+
+    @expose("/<pk>/permalink", methods=["POST"])
+    @protect()
+    @safe
+    @event_logger.log_this_with_context(
+        action=lambda self, *args, **kwargs: f"{self.__class__.__name__}.post",
+        log_to_statsd=False,
+    )
+    @requires_json
+    def post(self, pk: str) -> Response:
+        """Stores a new permanent link.
+        ---
+        post:
+          description: >-
+            Stores a new permanent link.
+          parameters:
+          - in: path
+            schema:
+              type: string
+            name: pk
+          requestBody:
+            required: true
+            content:
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/DashboardPermalinkPostSchema'
+          responses:
+            201:
+              description: The permanent link was stored successfully.
+              content:
+                application/json:
+                  schema:
+                    type: object
+                    properties:
+                      key:
+                        type: string
+                        description: The key to retrieve the permanent link data.
+                      url:
+                        type: string
+                        description: permanent link.
+            400:
+              $ref: '#/components/responses/400'
+            401:
+              $ref: '#/components/responses/401'
+            422:
+              $ref: '#/components/responses/422'
+            500:
+              $ref: '#/components/responses/500'
+        """
+        key_type = current_app.config["PERMALINK_KEY_TYPE"]
```

Comment:
I don't think we should provide this option. If security risk with numeric values is real, then it's real for everyone. If not, then we should just not worry about it.

An actual option I'd like to see is the ability to config the key length, which may require changing changing the keys from uuid (which only uses 16 hex characters) to case-sensitive [hashids](https://hashids.org/) used in real short-url services (e.g., bit.ly and t.co).

Reply:
I added this option to per a request from @graceguo-supercat, see discussion here:  https://github.com/apache/superset/pull/18181#issuecomment-1056467837

Reply:
Having said that, I really like the readability of `hashid`s, so I'm happy to make that the default and only supported permalink key type, thus eliminating this config flag. @graceguo-supercat is that ok for you? I can open a follow-up PR to address this (should be fairly simple).

Reply:
The cache endpoint is already using [token_url](https://docs.python.org/3/library/secrets.html#secrets.token_urlsafe). Since it supports length configuration, can't we just use it with a smaller length instead of introducing another dependency?